syntax = "proto3";
package srv6pm;

message ColorOptions {
    uint32 interval_duration = 1;
    uint32 delay_margin = 2;
    uint32 numbers_of_color = 3;
}

message StopExperimentRequest {
    string sdlist = 1;
}

message RetriveExperimentDataRequest {
    string sdlist = 1;
}

message MeasurementData {
    uint32 meas_id = 1;
    uint32 interval = 2;
    string timestamp = 3;
    // forward path
    uint64 ssSeqNum = 4;
    uint64 ssTxCounter = 5;
    uint64 rfRxCounter = 6;
    uint32 fwColor = 7;
    // reverse path
    uint64 rfSeqNum = 8;
    uint64 rfTxCounter = 9;
    uint64 ssRxCounter = 10;
    uint32 rvColor = 11;
}

message ExperimentDataResponse {
    StatusCode status = 1;
    repeated MeasurementData measurement_data = 2;
}

message StopExperimentReply {
    StatusCode status = 1;
}

enum MeasurementProtocol {
    TWAMP = 0;
    STAMP = 1;
}

enum StatusCode {
  STATUS_SUCCESS = 0;
  STATUS_OPERATION_NOT_SUPPORTED = 1;
  STATUS_BAD_REQUEST = 2;
  STATUS_INTERNAL_ERROR = 3;
  STATUS_INVALID_GRPC_REQUEST = 4;
  STATUS_FILE_EXISTS = 5;
  STATUS_NO_SUCH_PROCESS = 6;
  STATUS_INVALID_ACTION = 7;
  STATUS_GRPC_SERVICE_UNAVAILABLE = 8;
  STATUS_GRPC_UNAUTHORIZED = 9;
  STATUS_NOT_CONFIGURED = 10;
}

enum AuthenticationMode {
    HMAC_SHA_256 = 0;
}

enum MeasurementType {
    DELAY = 0;
    LOSS = 1;
}

enum TimestampFormat {
    PTPv2 = 0;
    NTP = 1;
}

enum MeasurementDelayMode {
    OneWay = 0;
    TwoWay = 1;
    LoopbackMode = 2;
}

enum MeasurementLossMode {
    Inferred = 0;
    Direct = 1;
}

enum PMDriver {
    eBPF = 0;
    IPSet = 1;
}

enum Configuration {
    uint32 ss_udp_port = 1;
    uint32 refl_udp_port = 2;
    ColorOptions color_options = 3;
    PMDriver pm_driver = 4;
}

message SetConfigurationReply {
    StatusCode status = 1;
}